cmd/geth: added counters to the geth inspect report

Neurone opened the pull request:
Add counters to the stats from inspect command.

These counters are useful to better understand what's inside LevelDB and they can be useful to better answer some recurrent questions about geth stats like _how many state entries I must download before having a synced full node_? (see https://github.com/ethereum/go-ethereum/issues/14647)

This is a preview of the output, based on a goerli synced full node:

```zsh
❯ ./build/bin/geth --goerli inspect
INFO [08-27|13:51:50.845] Maximum peer count                       ETH=50 LES=0 total=50
INFO [08-27|13:51:50.845] Smartcard socket not found, disabling    err="stat /run/pcscd/pcscd.comm: no such file or directory"
INFO [08-27|13:51:50.848] Set global gas cap                       cap=25000000
INFO [08-27|13:51:50.848] Allocated cache and file handles         database=/home/developer/.ethereum/goerli/geth/chaindata cache=512.00MiB handles=524288
INFO [08-27|13:51:51.094] Opened ancient database                  database=/home/developer/.ethereum/goerli/geth/chaindata/ancient
INFO [08-27|13:51:51.110] Persisted trie from memory database      nodes=361 size=51.17KiB time="866.628µs" gcnodes=0 gcsize=0.00B gctime=0s livenodes=1 livesize=0.00B
INFO [08-27|13:51:51.112] Loaded most recent local header          number=3297447 hash="a32882…d2974c" td=4812390 age=11s
INFO [08-27|13:51:51.112] Loaded most recent local full block      number=3297447 hash="a32882…d2974c" td=4812390 age=11s
INFO [08-27|13:51:51.112] Loaded most recent local fast block      number=3297447 hash="a32882…d2974c" td=4812390 age=11s
INFO [08-27|13:51:51.112] Loaded last fast-sync pivot marker       number=3292045
INFO [08-27|13:51:59.134] Inspecting database                      count=12630000 elapsed=8.000s
INFO [08-27|13:52:07.134] Inspecting database                      count=24366000 elapsed=16.000s
+-----------------+--------------------+------------+----------+
|    DATABASE     |      CATEGORY      |    SIZE    |  COUNT   |
+-----------------+--------------------+------------+----------+
| Key-Value store | Headers            | 13.21 MiB  |    21399 |
| Key-Value store | Bodies             | 40.74 MiB  |    21572 |
| Key-Value store | Receipts           | 37.35 MiB  |    21572 |
| Key-Value store | Difficulties       | 1.03 MiB   |    22038 |
| Key-Value store | Block number->hash | 948.25 KiB |    21131 |
| Key-Value store | Block hash->number | 128.97 MiB |  3298371 |
| Key-Value store | Transaction index  | 10.44 MiB  |   304051 |
| Key-Value store | Bloombit index     | 102.89 MiB |  1646592 |
| Key-Value store | Contract codes     | 97.26 MiB  |    17962 |
| Key-Value store | Trie nodes         | 2.37 GiB   | 24296985 |
| Key-Value store | Trie preimages     | 3.90 MiB   |    54925 |
| Key-Value store | Account snapshot   | 0.00 B     |        0 |
| Key-Value store | Storage snapshot   | 0.00 B     |        0 |
| Key-Value store | Clique snapshots   | 80.15 KiB  |       99 |
| Key-Value store | Singleton metadata | 151.00 B   |        5 |
| Ancient store   | Headers            | 1.04 GiB   | NA       |
| Ancient store   | Bodies             | 1.09 GiB   | NA       |
| Ancient store   | Receipts           | 596.63 MiB | NA       |
| Ancient store   | Difficulties       | 31.21 MiB  | NA       |
| Ancient store   | Block number->hash | 118.75 MiB | NA       |
| Light client    | CHT trie nodes     | 0.00 B     |        0 |
| Light client    | Bloom trie nodes   | 0.00 B     |        0 |
+-----------------+--------------------+------------+----------+
|                         TOTAL        |  5.66 GIB  |           
+-----------------+--------------------+------------+----------+
ERROR[08-27|13:52:10.258] Database contains unaccounted data       size=100.74KiB count=3514
```

holiman commented:
For the answer to "how many states", we do have a meter showing how much has been imported. 

![Screenshot_2020-08-27 Dual Geth - Grafana(2)](https://user-images.githubusercontent.com/142290/91440366-c822ec80-e86e-11ea-9a95-459cffb630f7.png)

We also have a statecounter in the db, that answers the same question, and that's why `imported` doesn't restart from zero if you stop and restart the sync. 

So that particular motivator is a bit moot, I think, particularly since if the number from this PR is used, is going to over-estimate the true number. A fast-synced node is near-optimal in terms of pruning, whereas the longer you run it after the pivot point, the more 'useless' states assemble (because we don't do perfect pruning). 

But that point aside, I think the feature in this PR is good to have for other reasons.

holiman commented on core/rawdb/database.go:
```diff
@@ -350,37 +393,37 @@ func InspectDatabase(db ethdb.Database) error {
+		{"Key-Value store", "Storage snapshot", storageSnapSize.String(), formatCounter(storageSnapSizeCount)},
+		{"Key-Value store", "Clique snapshots", cliqueSnapsSize.String(), formatCounter(cliqueSnapsSizeCount)},
+		{"Key-Value store", "Singleton metadata", metadata.String(), formatCounter(metadataCount)},
+		{"Ancient store", "Headers", ancientHeaders.String(), "NA"},
```
These should be fairly easy to obtain too, by calling `Ancients()` above where it currently only calls `AncientSize()`. Needs only be done once for the whole group

Neurone replied:
I see, nice! I'll work on it.

holiman reviewed (commented):
Generally lgtm!

Neurone commented:
I updated the PR:
- added stats for ancient db
- receipts are now counted even in ancient db
- cleaned up code using a struct instead of two variables for size and count

The output is now something like this:

```zsh
❯ ./build/bin/geth --goerli inspect
INFO [08-28|03:52:11.860] Maximum peer count                       ETH=50 LES=0 total=50
INFO [08-28|03:52:11.860] Smartcard socket not found, disabling    err="stat /run/pcscd/pcscd.comm: no such file or directory"
INFO [08-28|03:52:11.862] Set global gas cap                       cap=25000000
INFO [08-28|03:52:11.862] Allocated cache and file handles         database=/home/developer/.ethereum/goerli/geth/chaindata cache=512.00MiB handles=524288
INFO [08-28|03:52:11.933] Opened ancient database                  database=/home/developer/.ethereum/goerli/geth/chaindata/ancient
INFO [08-28|03:52:11.943] Persisted trie from memory database      nodes=361 size=51.17KiB time=1.366622ms gcnodes=0 gcsize=0.00B gctime=0s livenodes=1 livesize=0.00B
INFO [08-28|03:52:11.946] Loaded most recent local header          number=3300722 hash="2593ea…3b2c0f" td=4816955 age=21m42s
INFO [08-28|03:52:11.946] Loaded most recent local full block      number=3300722 hash="2593ea…3b2c0f" td=4816955 age=21m42s
INFO [08-28|03:52:11.946] Loaded most recent local fast block      number=3300722 hash="2593ea…3b2c0f" td=4816955 age=21m42s
INFO [08-28|03:52:11.947] Loaded last fast-sync pivot marker       number=3292045
INFO [08-28|03:52:19.999] Inspecting database                      count=16582000 elapsed=8.000s
INFO [08-28|03:52:28.003] Inspecting ancient database              number=681000  elapsed=16.004s
INFO [08-28|03:52:36.013] Inspecting ancient database              number=1721000 elapsed=24.014s
INFO [08-28|03:52:44.014] Inspecting ancient database              number=2126000 elapsed=32.015s
INFO [08-28|03:52:52.023] Inspecting ancient database              number=2537000 elapsed=40.024s
INFO [08-28|03:53:00.038] Inspecting ancient database              number=2880000 elapsed=48.039s
INFO [08-28|03:53:08.043] Inspecting ancient database              number=3120000 elapsed=56.044s
INFO [08-28|03:53:16.060] Inspecting ancient database              number=3256000 elapsed=1m4.061s
+-----------------+--------------------+-------------+----------+
|    DATABASE     |      CATEGORY      |    SIZE     |  COUNT   |
+-----------------+--------------------+-------------+----------+
| Key-Value store | Headers            | 15.35 MiB   |    24862 |
| Key-Value store | Bodies             | 46.06 MiB   |    24862 |
| Key-Value store | Receipts           | 40.59 MiB   |    84822 |
| Key-Value store | Difficulties       | 1.13 MiB    |    25253 |
| Key-Value store | Block number->hash | 1022.52 KiB |    24311 |
| Key-Value store | Block hash->number | 129.10 MiB  |  3301661 |
| Key-Value store | Transaction index  | 10.78 MiB   |   314049 |
| Key-Value store | Bloombit index     | 103.06 MiB  |  1648640 |
| Key-Value store | Contract codes     | 97.42 MiB   |    18008 |
| Key-Value store | Trie nodes         | 2.40 GiB    | 24443208 |
| Key-Value store | Trie preimages     | 7.87 MiB    |   110516 |
| Key-Value store | Account snapshot   | 0.00 B      |        0 |
| Key-Value store | Storage snapshot   | 0.00 B      |        0 |
| Key-Value store | Clique snapshots   | 82.61 KiB   |      102 |
| Key-Value store | Singleton metadata | 0.00 B      |        0 |
| Ancient store   | Headers            | 1.04 GiB    |  3276800 |
| Ancient store   | Bodies             | 1.09 GiB    |  3276800 |
| Ancient store   | Receipts           | 596.63 MiB  |  3761928 |
| Ancient store   | Difficulties       | 31.21 MiB   |  3276800 |
| Ancient store   | Block number->hash | 118.75 MiB  |  3276800 |
| Light client    | CHT trie nodes     | 0.00 B      |        0 |
| Light client    | Bloom trie nodes   | 0.00 B      |        0 |
+-----------------+--------------------+-------------+----------+
|                         TOTAL        |  5.70 GIB   |          |
+-----------------+--------------------+-------------+----------+
INFO [08-28|03:53:17.714] Inspection completed                     elapsed=1m5.715s
```

rjl493456442 commented on core/rawdb/database.go:
```diff
@@ -341,46 +382,67 @@ func InspectDatabase(db ethdb.Database) error {
+			ancientReceipts += countReceiptsRLP(data)
+		}
+		if blockNumber%1000 == 0 && time.Since(logged) > 8*time.Second {
+			log.Info("Inspecting ancient database", "number", blockNumber, "elapsed", common.PrettyDuration(time.Since(start)))
```
Would be nice to print like `Inspecting ancient database for receipts counting`.
And also we can print the total blocks needs to be iterated so that it's kind of progress marker

Neurone replied:
I created a percentage to have a better feeling of the progress. For the message, `Inspecting ancient database for receipts counting` goes over the formatting of other messages:

```zsh
INFO [08-28|05:23:49.768] Loaded most recent local header          number=3301083 hash="899a9a…3d2961" td=4817450 age=23m4s
INFO [08-28|05:23:49.768] Loaded most recent local full block      number=3301083 hash="899a9a…3d2961" td=4817450 age=23m4s
INFO [08-28|05:23:49.768] Loaded most recent local fast block      number=3301083 hash="899a9a…3d2961" td=4817450 age=23m4s
INFO [08-28|05:23:49.768] Loaded last fast-sync pivot marker       number=3300981
INFO [08-28|05:23:57.834] Inspecting database                      count=19601000 elapsed=8.000s
INFO [08-28|05:24:05.834] Inspecting ancient database for receipts counting number=373000 percentage=11 elapsed=16.000s
INFO [08-28|05:24:13.842] Inspecting ancient database for receipts counting number=1571000 percentage=47 elapsed=24.007s
INFO [08-28|05:24:21.845] Inspecting ancient database for receipts counting number=2042000 percentage=62 elapsed=32.010s
INFO [08-28|05:24:29.860] Inspecting ancient database for receipts counting number=2448000 percentage=74 elapsed=40.025s
```

What do you think of `Counting ancient database receipts`? Does it make sense in english?

```zsh
INFO [08-28|05:32:53.088] Loaded most recent local header          number=3301083 hash="899a9a…3d2961" td=4817450 age=32m8s
INFO [08-28|05:32:53.088] Loaded most recent local full block      number=3301083 hash="899a9a…3d2961" td=4817450 age=32m8s
INFO [08-28|05:32:53.088] Loaded most recent local fast block      number=3301083 hash="899a9a…3d2961" td=4817450 age=32m8s
INFO [08-28|05:32:53.088] Loaded last fast-sync pivot marker       number=3300981
INFO [08-28|05:33:01.199] Inspecting database                      count=19700000 elapsed=8.044s
INFO [08-28|05:33:09.206] Counting ancient database receipts       number=628000  percentage=19 elapsed=16.051s
INFO [08-28|05:33:17.223] Counting ancient database receipts       number=1694000 percentage=51 elapsed=24.068s
INFO [08-28|05:33:25.227] Counting ancient database receipts       number=2108000 percentage=64 elapsed=32.072s
INFO [08-28|05:33:33.232] Counting ancient database receipts       number=2529000 percentage=77 elapsed=40.078s
INFO [08-28|05:33:41.243] Counting ancient database receipts       number=2865000 percentage=87 elapsed=48.088s
INFO [08-28|05:33:49.257] Counting ancient database receipts       number=3111000 percentage=94 elapsed=56.103s
```

rjl493456442 replied:
Yes, it looks good to me. Maybe it would be better to add the 'final' state: `Counting ancient database receipts       number=3111000 percentage=100 elapsed=56.103s`

Neurone replied:
Done.

rjl493456442 commented on core/rawdb/database.go:
```diff
@@ -238,6 +240,43 @@ func NewLevelDBDatabaseWithFreezer(file string, cache int, handles int, freezer
+func countReceiptsRLP(data rlp.RawValue) counter {
+	// Convert the receipts from their storage form to their internal representation
+	storageReceipts := []*types.ReceiptForStorage{}
+	if err := rlp.DecodeBytes(data, &storageReceipts); err != nil {
```
RLP.Decode all the receipts in the database is expensive I think. Maybe we can apply some additional tricks here.
In RLP, we can count the elements in the list instead of decoding all bytes. You can check the example here https://github.com/ethereum/go-ethereum/blob/master/core/rawdb/chain_iterator.go#L140

cc @holiman Do you agree with me that using the RLP tricks here for receipts counting?

Neurone replied:
Man, it's very much faster! :racehorse: :racehorse: :racehorse: 

Before:

`INFO [08-28|05:33:59.162] Inspection completed                     elapsed=1m6.007s`

After:

`INFO [08-28|05:50:24.877] Inspection completed                     elapsed=20.951s`

rjl493456442 reviewed (commented):
I really like this PR, a few nitpicks.

rjl493456442 commented:
@Neurone How do you run the Geth? I notice that `transaction lookups` are away less than the total receipts.
Do you enable the transaction indexing feature by `txlookuplimit`?

Neurone commented:
> @Neurone How do you run the Geth? I notice that `transaction lookups` are away less than the total receipts.
> Do you enable the transaction indexing feature by `txlookuplimit`?

Long story short: I noticed too so now i deleted my state and resynced and all is good :)

There are some cases where you end up in that situation, in particular I currently found two both related to the rebuild from the ancient db.
- you delete the state db and not the ancient. Then you start directly `inspect`, then start geth normally. Because the first start was with inspect the maintainTxIndex function does not start, so you don't get old txs indexed anymore and you start indexing only new synced blocks
- similar situation: you delete the state db and not the ancient. You start geth normally and before maintainTxIndex has finished close geth. Again, the next start the maintainTxIndex does not start anymore and you end up in that situation.

I think it will be very useful to have a separate command to rebuild all indexes as you wish and in cases like these.

Neurone commented:
Here the current updated output of this PR using a Goerli full node:

```zsh
❯ ./build/bin/geth --goerli inspect
INFO [08-28|06:12:46.100] Maximum peer count                       ETH=50 LES=0 total=50
INFO [08-28|06:12:46.100] Smartcard socket not found, disabling    err="stat /run/pcscd/pcscd.comm: no such file or directory"
INFO [08-28|06:12:46.103] Set global gas cap                       cap=25000000
INFO [08-28|06:12:46.103] Allocated cache and file handles         database=/home/developer/.ethereum/goerli/geth/chaindata cache=512.00MiB handles=524288
INFO [08-28|06:12:46.149] Opened ancient database                  database=/home/developer/.ethereum/goerli/geth/chaindata/ancient
INFO [08-28|06:12:46.157] Persisted trie from memory database      nodes=361 size=51.17KiB time="863.5µs" gcnodes=0 gcsize=0.00B gctime=0s livenodes=1 livesize=0.00B
INFO [08-28|06:12:46.159] Loaded most recent local header          number=3301083 hash="899a9a…3d2961" td=4817450 age=1h12m1s
INFO [08-28|06:12:46.159] Loaded most recent local full block      number=3301083 hash="899a9a…3d2961" td=4817450 age=1h12m1s
INFO [08-28|06:12:46.159] Loaded most recent local fast block      number=3301083 hash="899a9a…3d2961" td=4817450 age=1h12m1s
INFO [08-28|06:12:46.159] Loaded last fast-sync pivot marker       number=3300981
INFO [08-28|06:12:54.225] Inspecting database                      count=22050000 elapsed=8.000s
INFO [08-28|06:13:02.227] Counting ancient database receipts       number=1798000 percentage=54 elapsed=16.001s
+-----------------+--------------------+------------+----------+
|    DATABASE     |      CATEGORY      |    SIZE    |  COUNT   |
+-----------------+--------------------+------------+----------+
| Key-Value store | Headers            | 15.00 MiB  |    24295 |
| Key-Value store | Bodies             | 44.52 MiB  |    24295 |
| Key-Value store | Receipts           | 39.90 MiB  |    83630 |
| Key-Value store | Difficulties       | 1.10 MiB   |    24681 |
| Key-Value store | Block number->hash | 1.01 MiB   |    24666 |
| Key-Value store | Block hash->number | 129.07 MiB |  3301094 |
| Key-Value store | Transaction index  | 132.03 MiB |  3845543 |
| Key-Value store | Bloombit index     | 103.06 MiB |  1648640 |
| Key-Value store | Contract codes     | 97.48 MiB  |    18032 |
| Key-Value store | Trie nodes         | 2.33 GiB   | 24204339 |
| Key-Value store | Trie preimages     | 109.35 KiB |     1548 |
| Key-Value store | Account snapshot   | 0.00 B     |        0 |
| Key-Value store | Storage snapshot   | 0.00 B     |        0 |
| Key-Value store | Clique snapshots   | 20.41 KiB  |       26 |
| Key-Value store | Singleton metadata | 0.00 B     |        0 |
| Ancient store   | Headers            | 1.04 GiB   |  3276800 |
| Ancient store   | Bodies             | 1.09 GiB   |  3276800 |
| Ancient store   | Receipts           | 596.63 MiB |  3761928 |
| Ancient store   | Difficulties       | 31.21 MiB  |  3276800 |
| Ancient store   | Block number->hash | 118.75 MiB |  3276800 |
| Light client    | CHT trie nodes     | 0.00 B     |        0 |
| Light client    | Bloom trie nodes   | 0.00 B     |        0 |
+-----------------+--------------------+------------+----------+
|                         TOTAL        |  5.74 GIB  |          |
+-----------------+--------------------+------------+----------+
INFO [08-28|06:13:06.521] Inspection completed                     elapsed=20.295s
```

holiman commented on core/rawdb/database.go:
```diff
@@ -341,46 +388,67 @@ func InspectDatabase(db ethdb.Database) error {
+	}
+	// Count receipts in ancient db
+	ancientReceipts := counter(0)
+	for blockNumber := uint64(0); blockNumber < uint64(ancients); blockNumber++ {
+		data, err := db.Ancient(freezerReceiptTable, blockNumber)
+		if err != nil {
+			log.Error("Error reading ancient receipts from block", "number", blockNumber, "err", err)
+		} else {
+			ancientReceipts += countReceiptsRLP(data)
+		}
+		if blockNumber%1000 == 0 && time.Since(logged) > 8*time.Second {
+			log.Info("Counting ancient database receipts", "number", blockNumber, "percentage", ancients.Percentage(blockNumber), "elapsed", common.PrettyDuration(time.Since(start)))
+			logged = time.Now()
+		}
```
Hm... I'm not fond of this.. So, in the 'ideal' case, all ancients are kept in lock-step, so we have identical number of headers, receipts etc. 
In this case, you've found that it's not the case, and when one does a `geth inspect`, it's good that we don't just trust the 'ideal case' but actually verify. However, 

1. if we _do_ verify, I don't see why we should treat receipts as a special snowflake, but do the same of all of them, 
2. There is a better way to check the items instead of reading all the data from disk, namely to expose `t.items` from the `freezerTable`, as maybe `Count()`

Neurone replied:
I think I'm not explained well. We found discrepancies in the tx index for the levelDB store, but ancient store is fine and we didn't find any discrepancy in numbers of headers, bodies, etc. vs receipts.

The PR uses `db.Ancients()` already to count elements from freezer, and the results correspond exactly to the number of blocks saved in the freezer. But - unlikely with levelDB - in freezer receipts are packed in blocks. Because info about levelDB and freezer are in the same report, that count can be misleading to user.

This is why in my opinion _Receipts_ value should be treated differently from the others, because usually we see them as single elements and not as packet of receipts, and because in the key-value store stat we gave just few line above they are counted as separated items.

At the start the following would have been the result based only on `db.Ancients()`

```zsh
+-----------------+--------------------+------------+----------+
|    DATABASE     |      CATEGORY      |    SIZE    |  COUNT   |
+-----------------+--------------------+------------+----------+
| Key-Value store | Receipts           | 39.90 MiB  |    83630 |
...
| Ancient store   | Bodies             | 1.09 GiB   |  3276800 |
| Ancient store   | Receipts           | 596.63 MiB |  3276800 |
| Ancient store   | Difficulties       | 31.21 MiB  |  3276800 |
...
```

But that _Receipts_ count seems misleading to me. Changing the label to _Recepit Blocks_ would have done the trick, but it would have not add much information to how many receipts there are in the ancient, so from there the choice to count receipts with an ad hoc procedure.

Maybe we can opt to use a separate value to preserve that info also, but I don't like very much this method because size column sounds misleading in every case:

```zsh
+-----------------+--------------------+------------+----------+
|    DATABASE     |      CATEGORY      |    SIZE    |  COUNT   |
+-----------------+--------------------+------------+----------+
| Key-Value store | Receipts           | 39.90 MiB  |    83630 |
...
| Ancient store   | Receipt Blocks     | 0.0 B      |  3276800 |
| Ancient store   | Receipts           | 596.63 MiB |  3761928 |
```
```zsh
+-----------------+--------------------+------------+----------+
|    DATABASE     |      CATEGORY      |    SIZE    |  COUNT   |
+-----------------+--------------------+------------+----------+
| Key-Value store | Receipts           | 39.90 MiB  |    83630 |
...
| Ancient store   | Receipt Blocks     | 596.63 MiB |  3276800 |
| Ancient store   | Receipts           | 596.63 MiB |  3761928 |
```

What do you think, does all this make sense for you?

holiman replied:
I think it makes more sense to change the label, so that for leveldb keep `Receipts`, as in "individual per-transaction receipts", but change `Ancient store | Receipt lists| size | number of items | ` (or alternatively `Receipt Blocks`). 

Quirky sidenote: `3845543-83630-3761928 = -15` == `(number of transactiions indexed) - (number of receipts in leveldb) - (number of receipts in anceintdb) == -15`. I guess there's some data which overlaps and is present both in ancients and leveldb when you ran it.

Neurone replied:
> I think it makes more sense to change the label, so that for leveldb keep `Receipts`, as in "individual per-transaction receipts", but change `Ancient store | Receipt lists| size | number of items | ` (or alternatively `Receipt Blocks`).

Ok, but this way you don't know how many receipts are really stored in ancient db, so you cannot easily do consistency analysis like you did below :) What if we put both info in the report, something like this (changed column title and added counted receipts)?

```zsh
+-----------------+--------------------+------------+----------+
|    DATABASE     |      CATEGORY      |    SIZE    |  ITEMS   |
+-----------------+--------------------+------------+----------+
| Key-Value store | Headers            | 15.00 MiB  |    24295 |
| Key-Value store | Bodies             | 44.52 MiB  |    24295 |
| Key-Value store | Receipts           | 39.90 MiB  |    83630 |
...
| Ancient store   | Receipt lists      | 596.63 MiB |  3276800 |
| Ancient store   | └ counted receipts | --         |  3761928 |
| Ancient store   | Difficulties       | 31.21 MiB  |  3276800 |
...
```

> Quirky sidenote: `3845543-83630-3761928 = -15` == `(number of transactiions indexed) - (number of receipts in leveldb) - (number of receipts in anceintdb) == -15`. I guess there's some data which overlaps and is present both in ancients and leveldb when you ran it.

Interesting. Maybe during chain reorgs the tx index is cleaned up but old receipts are kept in db anyway? I'll try to understand.

Neurone commented:
Hi, is there something more I can do on this PR or is it ok?

holiman commented:
I stil don't like treating `receiptBlocks` as special. We also have block bodies, `block`s, in ancient db, which consist of lists of transactions and uncle-headers. So if we go down the "detailed inspection of freezer" path, then those should be counted too. 

All in all, I don't think we should do that, but just count the `items' (blobs) in freezer. However, that's my personal pref, maybe @karalabe or @rjl493456442 feels differently?

Neurone commented:
I see your point. What if we introduce a `deepinspect` flag to activate this kind of detailed inspections? I can remove the receipts count from this PR, wait for merge, then submit a  new PR with the new flag, the receipts count 
and some other deep inspection. What do you think?

Neurone commented:
It seems there aren't other feedbacks, so I removed the counting of receipts in ancient store. I hope this PR sounds fine now.

rjl493456442 commented on core/rawdb/database.go:
```diff
@@ -238,6 +239,50 @@ func NewLevelDBDatabaseWithFreezer(file string, cache int, handles int, freezer
+		log.Warn("Receipt iteration error", "error", err)
+		return counter(0)
+	}
+	count := counter(0)
```
Can we use the List counter of RLP? Instead of iterating the whole list.

holiman reviewed (approved):
LGTM

rjl493456442 commented on core/rawdb/database.go:
```diff
@@ -289,98 +336,105 @@ func InspectDatabase(db ethdb.Database) error {
+	ancients := counter(0)
+	if count, err := db.Ancients(); err == nil {
+		ancients = counter(count)
+	}
```
If you already count the receipts number in the leveldb, why not also counting them in the ancient store?

holiman commented on core/rawdb/database.go:
```diff
@@ -289,98 +336,105 @@ func InspectDatabase(db ethdb.Database) error {
+			bodies.Add(size)
 		case bytes.HasPrefix(key, blockReceiptsPrefix) && len(key) == (len(blockReceiptsPrefix)+8+common.HashLength):
-			receiptSize += size
+			receipts.size += size
+			receipts.count += countReceiptsRLP(it.Value())
```
Ah, I missed this one. So here you are counting individual receipts. But if we just removed "individual receipts counting" from ancientdb, it doesn't make much sense to count them individually in leveldb. Maybe just report "receipt lists" in leveldb aswell

Neurone replied:
Done. I'll come back proposing those counters someday though ... :D Just to be sure before starting to work on: what do you think if in a next PR I add something like a "deepinspect" flag and I do other inspections and counting, and not onlyabout  receipts?

holiman commented:
Deepinspect tag sounds good to me!

holiman commented:
Thanks for working with us on this one!
